test/apps/iodemo/io_demo.cc

```diff
@@ -584,68 +633,96 @@ class DemoClient : public P2pDemoCommon {
     }
 
     bool run() {
-        std::vector<UcxConnection*> conn;
-        conn.resize(opts().servers.size());
-        for (size_t i = 0; i < conn.size(); i++) {
-            conn[i] = connect(opts().servers[i]);
-            if (!conn[i]) {
-                LOG << "Connect to server ["
-                    << opts().servers[i]
-                    << "] Failed!";
-                for (size_t j = 0; j < i; j++) {
-                    delete conn[j];
+        if (_server_info.empty()) {
+            _server_info.resize(opts().servers.size());
+        }
+        for (size_t i = 0; i < _server_info.size(); i++) {
+            if (_server_info[i].conn == NULL) {
+                _server_info[i].conn = connect(opts().servers[i], i + 1);
+                if (!_server_info[i].conn) {
+                    LOG << "Connect to server ["
+                        << opts().servers[i]
+                        << "] Failed!";
+                } else {
+                    _num_connected++;
                 }
-                return false;
             }
         }
-
+        if (_num_connected == 0) {
+            return false;
+        }
         _status = OK;
 
         // TODO reset these values by canceling requests
         _num_sent      = 0;
         _num_completed = 0;
 
-        double prev_time     = get_time();
-        long total_iter      = 0;
-        long total_prev_iter = 0;
+        double prev_time           = get_time();
+        double prev_reconnect_time = get_time();
+        long total_iter            = 0;
+        long total_prev_iter       = 0;
         std::vector<op_info_t> info;
 
         for (int i = 0; i < IO_COMP; ++i) {
             op_info_t op_info = {static_cast<io_op_t>(i), 0, 0};
             info.push_back(op_info);
         }
 
-        while ((total_iter < opts().iter_count) && (_status == OK)) {
+        while ((total_iter < opts().iter_count)) {
             VERBOSE_LOG << " <<<< iteration " << total_iter << " >>>>";
 
             if (!wait_for_responses(opts().window_size - 1)) {
                 break;
             }
 
-            size_t conn_num = IoDemoRandom::rand(0, conn.size() - 1);
-            io_op_t op      = get_op();
+            if (_num_connected < _server_info.size()) {
+                double curr_time = get_time();
+                if (curr_time >= (prev_reconnect_time + 3.0)) {
+                    prev_reconnect_time = curr_time;
+                    for (size_t i = 0; i < _server_info.size(); i++) {
+                        if (_server_info[i].conn == NULL) {
+                            _server_info[i].conn = connect(opts().servers[i], i + 1);
```

Comment:
let's add `opts().servers[i]` address to `_server_info[i]` and pass `_server_info[i]` reference to `connect` instead of `opts().servers[i]`. So, we can access `_server_info` inside connection without indexing.